packages/angular_devkit/core/src/jobs/architecture.md

```diff
@@ -0,0 +1,253 @@
+
+# Glossary
+- `job handler`. The function that implements the job's logic.
+- `raw input`. The input observable sending messages to the job. These messages are of type
+    `JobInput`.
+- `raw output`. The output observer returned from the `job handler`. Messages on this observable
+    are of type `JobEvent`.
+
+# Description
+
+A `JobHandler`, similar to observables, is a function that receives an argument and a context, and
+returns an `Observable` of events, which can include outputs that are typed at runtime (using a
+Json Schema):
+
+```javascript
+const output1ToXEverySecond = function (x, context) {
+  return new Observable(subscriber => {
+    let t = 0;
+
+    // Notify our users that the actual work is started.
+    subscriber.next({ kind: JobEventKind.Start });
+    const i = setInterval(() => {
+      t++;
+      subscriber.next({ kind: JobEventKind.Output, value: t });
+      if (t === x) {
+        subscriber.next({ kind: JobEventKind.End });
+        subscriber.complete();
+      }
+    }, 1000);
+
+    return () => {
+      clearInterval(i);
+    };
+  })
+};
+
+// For now, jobs can not be called without a registry and scheduler.
+const registry = new SimpleJobRegistry();
+registry.register('output-from-1-to-x', output1ToXEverySecond, {
+  argument: { type: 'number' },
+  output: { type: 'number' },
+});
+const scheduler = new SimpleScheduler(registry);
+
+// Need to keep the same name that the registry would understand.
+// Count from 1 to 10.
+const job = scheduler.schedule('output-from-1-to-x', 10);
+
+// A Job<> instance has more members, but we only want the output values here.
+job.output.subscribe(x => console.log(x));
+```
+
+This seems like a lot of boilerplate in comparison, but there are a few advantages;
+
+1. lifecycle. Jobs can tell when they start doing work and when work is done.
+1. everything is typed, even at runtime.
+1. the context also contains an input Observable that receives typed input events, including
+    input values, and stop requests.
+1. jobs can also schedule other jobs and wait for them, even if they don't know if a job is
+    implemented in the system.
+
+## Diagram
+A simpler way to think about jobs in contrast to observables is that job are closer to a Unix
+process. It has an argument (command line flags), receive inputs (STDIN and interupt signals),
+and output values (STDOUT) as well as diagnostic (STDERR). They can be plugged one into another
+(piping), and can be transformed, synchronized and scheduled (fork, exec, cron).
+
+```plain
+- given A the type of the argument
+- given I the type of the input
+- given O the type of the output
+
+                              ,______________________   ,----> rawOuput: JobEvent<O>
+    rawInput: JobInput<I> --> | handler(argument: A) |            - JobEventKind.Log
+                              `⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻⎻             - JobEventKind.Output
+                                                                  - ...
```

Comment:
Something seems to have gone wrong in this diagram, I think you wanted the `----> rawOuput:` bit to be in the middle of the argument box (but maybe I am wrong).

Reply:
Nah I wanted to have it over top as I had other arrows but reduced to one.